WIP yaml based-patch system because this repo broke me

beckermr opened the pull request:
Checklist
* [ ] Ran `python show_diff.py` and posted the output as part of the PR.
* [ ] Modifications won't affect packages built in the future. <!-- Make sure to add a condition `and record.get("timestamp", 0) < NOW` so your changes only affect packages built in the past. Replace NOW with `python -c "import time; print(f'{time.time():.0f}000')"` -->

beckermr commented on recipe/meta.yaml:
```diff
@@ -30,6 +30,7 @@ test:
     - gen_patch_json.py
     - get_license_family.py
     - show_diff.py
+    patch_yamls_utils.py
```
```suggestion
    - patch_yamls_utils.py
```

beckermr commented:
@conda-forge-admin rerender

beckermr commented:
OK @conda-forge/core. I finally got so frustrated with this repo that I broke down and wrote a small YAML-based system for common patch patterns. I don't have a ton of more energy for this and would like to see this merged since we absolutely need it to manage the complexity here. 

So please add your comments, keeping in mind that I will very likely not make any large-scale changes to the code due to time. If you have strong feelings that this cannot be merged without a ton of extra work, then we can let this sit. 

I know the code needs tests. I'll write some basic tests once we have some semblance of agreement on the YAML format and the API. 

See the README.md in the recipe/patch_yaml directory for how it all works.

beckermr commented on recipe/patch_yaml/README.md:
```diff
@@ -0,0 +1,113 @@
+
+  # any key in the repodata entry (e.g., "version" or "build_number") with an operation
+  <repodata key>_<ge, gt, le, lt>: <value>
+
```
```suggestion
```

beckermr commented on recipe/patch_yaml/README.md:
```diff
@@ -0,0 +1,113 @@
+  build_number_in: [0, 1, 2]
+
+  # has specific dependencies as either a list or a single string
+  has_depends: numpy *  # matches any numpy entry with or without a version
```
```suggestion
  has_depends: numpy *  # matches any numpy entry with or without a version
  has_depends: numpy  # matches "numpy" exactly (i.e., no pins)
```

beckermr commented on recipe/meta.yaml:
```diff
@@ -30,25 +30,36 @@ test:
   home: https://github.com/conda-forge/conda-forge-repodata-patches-feedstock
   license: CC-PDDC
+  description: |
+    See the [README](recipe/patch_yaml/README.md) for how generate patches from static YAML files.
```
```suggestion
    See the [recipe/patch_yaml/README.md](recipe/patch_yaml/README.md) for how generate patches from static YAML files.
```

beckermr commented:
@conda-forge-admin rerender

jaimergp commented on recipe/gen_patch_json.py:
```diff
@@ -12,9 +12,20 @@
+from packaging.version import parse as parse_version
 
 from get_license_family import get_license_family
+from patch_yaml_utils import (
+    patch_yaml_edit_index,
+    _extract_track_feature,
+    _replace_pin,
+    _rename_dependency,
+    _relax_exact,
+    _pin_stricter,
+    _pin_looser,
+    CB_PIN_REGEX,
+    pad_list,
+)
```
Nice. Most of these were originally in this file but then moved to the utils module, right? I am not sure if that module should be named `patch_yaml_utils` though, since its functions are used in other stuff too.

beckermr replied:
My intent is that all patches end up as yaml and so the utility functions are first class and old used in a hacky way in the original code. Thus I think the name (and also the fact that these utils have `_` at the start still) is sensible to me.

jaimergp commented on recipe/patch_yaml/README.md:
```diff
@@ -0,0 +1,113 @@
+
+  # make a dependency version constraint stricter
+  - tighten_depends:
+      # package to pin stricter
```
To be clear, this is the package _name_ right? Not the full spec. I don't see the `old` vs `new` pairing here, so just double checking. I could have understood this as "gimme the full thing as you find it in the record", so maybe `name` or `pkg_name` would be clearer.

jaimergp commented on recipe/patch_yaml/README.md:
````diff
@@ -0,0 +1,113 @@
+- Multiple patches can be in the same file using separate YAML documents (i.e., separate the data by `---`
+  on a new line).
+
+```yaml
````
It would be nice to have this as `pydantic` model / JSON schema for easy validation before running the patch. I can contribute one if think it's a good idea!

beckermr replied:
Yes please feel free to contribute directly to this PR!

jaimergp commented on recipe/patch_yaml_utils.py:
```diff
@@ -0,0 +1,310 @@
+    for k, v in patch_yaml["if"].items():
+        if k == "subdir_in":
+            keep = keep and _fnmatch_str_or_list(subdir, v)
+            if not keep:
+                break
```
I think we can move this block to the end of the elif chain (after the else) instead of having it in every condition, given that they all are exclusive.

jaimergp commented on recipe/patch_yaml_utils.py:
```diff
@@ -0,0 +1,310 @@
+            record["depends"] = depends
+
+
+CB_PIN_REGEX = re.compile(r"^>=(?P<lower>\d(\.\d+)*a?),<(?P<upper>\d(\.\d+)*)a0$")
```
I know this is existing code that moved from the other module, but I wonder how tricky it'd be to use the `MatchSpec` here directly to avoid this kind of regex.

beckermr replied:
I am definitely going to punt on this one, but feel free to try!

beckermr replied:
Given how minimal our testing is here and how essential this data is, I am already nervous about big changes.

xhochy commented on recipe/patch_yaml/README.md:
```diff
@@ -0,0 +1,113 @@
+      old: matplotlib ==1.3.0
+      # thing to replace `old` with
+      new: matplotlib-base ==1.4.0
+      # ^^^^^^ NOTICE the extra indent here!
```
I'm confused by the comment. I would have expected `new` to be then indented one more than `old`.  Maybe skip the comment for clarity?

beckermr commented:
OK @jaimergp @xhochy 

I fixed a bug where some keys like timestamp are not in all record. I also addressed the code review where I could and made some of the code less verbose. 

I am going to write some tests since it appears the API is OK for now.

beckermr commented:
OK @jaimergp @xhochy This one is ready to go from my perspective. @jaimergp You can do the pydantic work here or in a separate PR, your call!
